### PR TITLE
Integration of Stepwise algorithm from otlm

### LINKED ISSUE #1209: LinearModelStepwiseAlgorithm from otlm does not exist in OT
According to the changes in the log :

https://github.com/openturns/openturns/blob/d15e64faef88e6647d8bc735f76bcb03a0b74819/ChangeLog#L77

new linear model features, which were previously available in otlm are now in OT, starting from 1.13. 

However, it seems that the 

https://github.com/openturns/otlm/blob/master/lib/src/LinearModelStepwiseAlgorithm.cxx

class does not belong to OT. 

I noticed that the otlm was not maintained anymore

http://trac.openturns.org/wiki/Modules?action=diff&version=62

which makes sense.

### PR DESCRIPTION
**Pull request by sofianehaddad**
The objective is to introduce the linear `stepwise` algorithm

We integrate the `otlm` algorithm as it is.

This fixes #1209

### CODE REVIEW

File: lib/src/Uncertainty/Algorithm/MetaModel/LinearModel/LinearModelStepwiseAlgorithm.cxx [hunk 1]
```diff
@@ -0,0 +1,684 @@
...
+
+    for (UnsignedInteger index = r.begin(); index != r.end(); ++index)
+    {
+      const UnsignedInteger i = indexSet_[index];
+      memcpy(&xi(0, 0), &Xmax_(0, i), sizeof(Scalar)*size);
+      const Matrix Qtxi(Q_.getImplementation()->genProd(*(xi.getImplementation()), true, false));
+      const Matrix di(Q_ * Qtxi);
+      const Matrix vi(xi - di);
+      memcpy(&viNP[0], &vi(0, 0), sizeof(Scalar)*size);
+      memcpy(&xiNP[0], &xi(0, 0), sizeof(Scalar)*size);
```
Comment (jschueller) [inline lib/src/Uncertainty/Algorithm/MetaModel/LinearModel/LinearModelStepwiseAlgorithm.cxx]: Prefer std::copy to memcpy

File: lib/src/Uncertainty/Algorithm/MetaModel/LinearModel/LinearModelStepwiseAlgorithm.cxx [hunk 2]
```diff
@@ -0,0 +1,684 @@
...
+  else
+  {
+    buildCurrentMatrixFromIndices(startIndices_);
+    for (UnsignedInteger i = 0; i < startIndices_.getSize(); ++i)
+      columnMaxToCurrent[startIndices_[i]] = i;
+  }
+  UnsignedInteger p = currentX_.getNbColumns();
+
+  UnsignedInteger iterations = 0;
+  Scalar Lstar;
```
Comment (jschueller) [inline lib/src/Uncertainty/Algorithm/MetaModel/LinearModel/LinearModelStepwiseAlgorithm.cxx]: Avoid unitialized variables

File: lib/src/Uncertainty/Algorithm/MetaModel/LinearModel/LinearModelStepwiseAlgorithm.cxx [hunk 3]
```diff
@@ -0,0 +1,684 @@
...
+  Point QtiNP(p);
+  for (UnsignedInteger i = 0; i < size; ++i)
+  {
+    memcpy(&QtiNP[0], &Qt(0, i), sizeof(Scalar)*p);
+    leverages[i] = QtiNP.dot(QtiNP);
+  }
+  Sample residualSample(size, 1);
+  memcpy(&residualSample(0, 0), &currentResidual_(0, 0), sizeof(Scalar) * size);
+
+  Point sigma2(residualSample.computeRawMoment(2));
```
Comment (jschueller) [inline lib/src/Uncertainty/Algorithm/MetaModel/LinearModel/LinearModelStepwiseAlgorithm.cxx]: const

File: lib/src/Uncertainty/Algorithm/MetaModel/LinearModel/LinearModelStepwiseAlgorithm.cxx [hunk 4]
```diff
@@ -0,0 +1,684 @@
...
+
+  Point diagonalGramInverse(p);
+  Point invRtiNP(p);
+  for (UnsignedInteger i = 0; i < p; ++i)
+  {
+    memcpy(&invRtiNP[0], &currentInvRt_(0, i), sizeof(Scalar)*p);
+    diagonalGramInverse[i] = invRtiNP.dot(invRtiNP);
+  }
+  Point leverages(size);
+  Matrix Qt(currentQ_.transpose());
```
Comment (jschueller) [inline lib/src/Uncertainty/Algorithm/MetaModel/LinearModel/LinearModelStepwiseAlgorithm.cxx]: const

File: python/src/LinearModelStepwiseAlgorithm_doc.i.in [hunk 5]
```diff
@@ -0,0 +1,65 @@
...
+    The indices of minimal model
+isForward : :class:`~openturns.Bool`
+      the boolean value used for the stepwise regression method direction FORWARD and BACKWARD. 
+startIndices : :class:`~openturns.Indices`
+     The indices of start model used for the stepwise regression method direction BOTH.
+penalty : float
+     The multiple of the degrees of freedom used for the penalty of the stepwise regression method : 
+     - 2      Akaike   information criterion (AIC)
+     - log(n) Bayesian information criterion (BIC)  
+maximumIterationNumber : :class:`~openturns.UnsignedInteger`
```
Comment (jschueller) [inline python/src/LinearModelStepwiseAlgorithm_doc.i.in]: for input types use bool, int, sequence of int, etc

**Comment by jschueller** [comment]
I dont remember what stopped this from being included before, otherwise looks ok.
It just misses the changelog update, mention the issue number it closes.

File: lib/src/Uncertainty/Algorithm/MetaModel/LinearModel/LinearModelStepwiseAlgorithm.cxx [hunk 6]
```diff
@@ -0,0 +1,684 @@
...
+
+CLASSNAMEINIT(LinearModelStepwiseAlgorithm)
+static const Factory<LinearModelStepwiseAlgorithm> Factory_LinearModelStepwiseAlgorithm;
+
+/* Default constructor */
+LinearModelStepwiseAlgorithm::LinearModelStepwiseAlgorithm()
+  : PersistentObject()
+  , direction_(FORWARD)
+  , penalty_(-1.0)
+  , maximumIterationNumber_(1000)
```
Comment (regislebrun) [inline lib/src/Uncertainty/Algorithm/MetaModel/LinearModel/LinearModelStepwiseAlgorithm.cxx]: A key in ResourceMap?

File: lib/src/Uncertainty/Algorithm/MetaModel/LinearModel/LinearModelStepwiseAlgorithm.cxx [hunk 7]
```diff
@@ -0,0 +1,684 @@
...
+      const Matrix Qtxi(Q_.getImplementation()->genProd(*(xi.getImplementation()), true, false));
+      const Matrix di(Q_ * Qtxi);
+      const Matrix vi(xi - di);
+      memcpy(&viNP[0], &vi(0, 0), sizeof(Scalar)*size);
+      memcpy(&xiNP[0], &xi(0, 0), sizeof(Scalar)*size);
+      const Scalar denominator = xiNP.dot(viNP);
+      if (denominator == 0.0) continue;
+      const Scalar alpha = xiNP.dot(residualNP) / denominator;
+      const Point newResidual(residualNP - alpha * viNP);
+      const Scalar newCriterion(newResidual.normSquare());
```
Comment (regislebrun) [inline lib/src/Uncertainty/Algorithm/MetaModel/LinearModel/LinearModelStepwiseAlgorithm.cxx]: `const Scalar newCriterion = newResidual.normSquare();`

File: lib/src/Uncertainty/Algorithm/MetaModel/LinearModel/LinearModelStepwiseAlgorithm.cxx [hunk 8]
```diff
@@ -0,0 +1,684 @@
...
+
+/* Perform regression */
+void LinearModelStepwiseAlgorithm::run()
+{
+  if (hasRun_) return;
+
+  const UnsignedInteger size(inputSample_.getSize());
+  Y_ = Matrix(size, 1, outputSample_.getImplementation()->getData());
+  const AggregatedFunction f(basis_);
+  Sample fx(f(inputSample_));
```
Comment (regislebrun) [inline lib/src/Uncertainty/Algorithm/MetaModel/LinearModel/LinearModelStepwiseAlgorithm.cxx]: const Sample

File: lib/src/Uncertainty/Algorithm/MetaModel/LinearModel/LinearModelStepwiseAlgorithm.cxx [hunk 9]
```diff
@@ -0,0 +1,684 @@
...
+  LOGDEBUG(OSS() << "Total number of columns=" << fx.getDimension());
+
+  {
+    // Reduce scope of Xt
+    Matrix Xt(fx.getDimension(), size, fx.getImplementation()->getData());
```
Comment (regislebrun) [inline lib/src/Uncertainty/Algorithm/MetaModel/LinearModel/LinearModelStepwiseAlgorithm.cxx]: const Matrix

File: lib/src/Uncertainty/Algorithm/MetaModel/LinearModel/openturns/LinearModelStepwiseAlgorithm.hxx [hunk 10]
```diff
@@ -0,0 +1,170 @@
...
+  LinearModelStepwiseAlgorithm();
+
+  /** Parameters constructor FORWARD and BACKWARD */
+  LinearModelStepwiseAlgorithm(const Sample & inputSample,
+                               const Basis & basis,
+                               const Sample & outputSample,
+                               const Indices & minimalIndices,
+                               const Bool isForward = true,
+                               const Scalar penalty = -1.0 /* < 0 means BIC, by convention */,
+                               const UnsignedInteger maximumIterationNumber = 1000);
```
Comment (regislebrun) [inline lib/src/Uncertainty/Algorithm/MetaModel/LinearModel/openturns/LinearModelStepwiseAlgorithm.hxx]: 1000 should be a key in ResourceMAp

File: lib/src/Uncertainty/Algorithm/MetaModel/LinearModel/openturns/LinearModelStepwiseAlgorithm.hxx [hunk 11]
```diff
@@ -0,0 +1,170 @@
...
+
+  /** Parameters constructor BOTH */
+  LinearModelStepwiseAlgorithm(const Sample & inputSample,
+                               const Basis & basis,
+                               const Sample & outputSample,
+                               const Indices & minimalIndices,
+                               const Indices & startIndices,
+                               const Scalar penalty = -1.0 /* < 0 means BIC, by convention */,
+                               const UnsignedInteger maximumIterationNumber = 1000);
```
Comment (regislebrun) [inline lib/src/Uncertainty/Algorithm/MetaModel/LinearModel/openturns/LinearModelStepwiseAlgorithm.hxx]: Here too

**Comment by sofianehaddad** [comment]
ok thanks guys I will take into account your feedbacks
@jschueller to answer your question, I had in mind to fully update this code. UnfortunatelyI didn't have time for and we need several API changes before to keep the current performance...

**Comment by jschueller** [comment]
you'll need to rebase, I let an error slip between merges

**Comment by sofianehaddad** [comment]
@jschueller because you didn't take into account my review on test ;-)

**Comment by jschueller** [comment]
It was an error related to Copula being removed in the system events PR

**Comment by sofianehaddad** [comment]
https://github.com/openturns/openturns/pull/1238/files#r333248630 and https://github.com/openturns/openturns/pull/1238/files#r333248201 ;-)

**Comment by jschueller** [comment]
sorry, I missed those

**Comment by jschueller** [comment]
It needs a last rebase :)

**Comment by sofianehaddad** [comment]
Done

File: python/test/t_LinearModelStepwiseAlgorithm_std.py [hunk 12]
```diff
@@ -0,0 +1,136 @@
+import openturns as ot
+from math import log
+
```
Comment (jschueller) [inline python/test/t_LinearModelStepwiseAlgorithm_std.py]: you need to call TESTPREABLE for the exponent formatting

**Comment by mbaudin47** [comment]
@sofianehaddad I cannot find the LinearModelStepwiseAlgorithm class in the doc:

LinearModelStepwiseAlgorithm

**Comment by jschueller** [comment]
done in #1310

**Pull request merged**